dali/base/sysinfologger.cpp

```diff
@@ -0,0 +1,682 @@
+            CSysInfoLoggerMsg & sysInfoMsg = iter->queryInfoLoggerMsg();
+            StringBuffer xpath;
+            sysInfoMsg.getXpath(xpath);
+            deleteXpathList.push_back(xpath.str());
+        }
+    }
+    IPropertyTree * root = conn->queryRoot();
+    unsigned count = 0;
+    for (auto & xpath: deleteXpathList)
+    {
+        if (root->removeProp(xpath.c_str()));
+            ++count;
+    }
+    return count;
+}
+
+bool deleteLogSysInfoMsg(unsigned __int64 msgId)
+{
+    Owned<ISysInfoLoggerMsgFilter> msgFilter = createSysInfoLoggerMsgFilter();
+    msgFilter->setMatchMsgId(msgId);
+    return deleteLogSysInfoMsg(msgFilter);
+}
+
+unsigned deleteOlderThanLogSysInfoMsg(bool visibleOnly, bool hiddenOnly, unsigned year, unsigned month, unsigned day)
+{
+    if (!year && month)
+        throw makeStringExceptionV(-1, "deleteOlderThanLogSysInfoMsg: year must be provided if month is specified (year=%u, month=%u, day=%u)", year, month, day);
+    if (!month && day)
+        throw makeStringExceptionV(-1, "deleteOlderThanLogSysInfoMsg: month must be provided if day is specified (year=%u, month=%u, day=%u)", year, month, day);
+    if (month>12)
+        throw makeStringExceptionV(-1, "deleteOlderThanLogSysInfoMsg: invalid month(year=%u, month=%u, day=%u)", year, month, day);
+    if (day>31)
+        throw makeStringExceptionV(-1, "deleteOlderThanLogSysInfoMsg: invalid day(year=%u, month=%u, day=%u)", year, month, day);
+    // With visibleOnly/hiddenOnly option, use createSysInfoLoggerMsgFilter()
+    if (visibleOnly || hiddenOnly)
+    {
+        unsigned count = 0;
+        Owned<ISysInfoLoggerMsgFilter> msgFilter = createSysInfoLoggerMsgFilter();
+        if (hiddenOnly)
+            msgFilter->setHiddenOnly();
+        if (visibleOnly)
+            msgFilter->setVisibleOnly();
+        msgFilter->setOlderThanDate(year, month, day);
+        return deleteLogSysInfoMsg(msgFilter);
+    }
+
+    // With only date range, use this quicker method to remove whole subtrees
+    Owned<IRemoteConnection> conn = querySDS().connect(SYS_INFO_ROOT, myProcessSession(), RTM_LOCK_WRITE, SDS_LOCK_TIMEOUT);
+    if (!conn)
+        return 0;
+
+    std::vector<std::string> deleteXpathList;
+    IPropertyTree * root = conn->queryRoot();
+    Owned<IPropertyTreeIterator> monthIter = root->getElements("*");
+    ForEach(*monthIter)
+    {
+        IPropertyTree & monthPT = monthIter->query();
+        if (year==0)
+            deleteXpathList.push_back(monthPT.queryName());
+        else
+        {
+            unsigned msgYear = 0, msgMonth = 0;
+            const char *p = monthPT.queryName(); // should be in format 'myyyydd'
+            if (*p++ == 'm')
+            {
+                msgYear = readDigits(p, 4, false);
+                msgMonth = readDigits(p, 2, false);
+            }
+            if (msgYear == 0 || msgMonth == 0)
+                throw makeStringExceptionV(-1, "child of " SYS_INFO_ROOT " is invalid: %s", monthPT.queryName());
```

Comment:
Should this log at the end of the function?  Otherwise an invalid node will prevent items from being deleted.